# Lesson 2: Data Input and Wrangling

## Introduction

In this lesson we will will learn to download a csv file from a website, modify it, and then use the file to download more files.

In [35]:
import pandas as pd
# import io
# import requests
# import re
# import logging
# import zipfile
from tqdm import tqdm


## 1 Getting Data

The most basic step in any data science project is actually getting the data. The pandas library makes this very easy if you are working with tabular data. Generally, you will be working with csv files and will be able to use the `read_csv` file.

We know from Gutenberg documentation that Gutenberg keeps a list of all of the books in its catalog a the following address: https://www.gutenberg.org/cache/epub/feeds/

From here we can download the `pg_catalog.csv.gz` file and unzip it.

This step has already been done for you, and the `pg_catalog.csv` file should be in the files you downloaded.


### 1.1 Loading a local file

To load the csv file locally, we only need to add run the command `pd.read_csv(<FILENAME>)`

In [36]:
pd.read_csv("pg_catalog.csv")

Text#  Type      Issued  \
0          1  Text  1971-12-01   
1          2  Text  1972-12-01   
2          3  Text  1973-11-01   
3          4  Text  1973-11-01   
4          5  Text  1975-12-01   
...      ...   ...         ...   
74250  74416  Text  2024-09-14   
74251  74417  Text  2024-09-14   
74252  74418  Text  2024-09-15   
74253  74419  Text  2024-09-15   
74254  90907  Text  2019-12-12   

                                                   Title Language  \
0      The Declaration of Independence of the United ...       en   
1      The United States Bill of Rights\r\nThe Ten Or...       en   
2                    John F. Kennedy's Inaugural Address       en   
3      Lincoln's Gettysburg Address\r\nGiven November...       en   
4                         The United States Constitution       en   
...                                                  ...      ...   
74250                                            Proteus       en   
74251                                   The kaleidoscope       en   
74252                                  Their island home       en   
74253                             Reuben Sachs; a sketch       en   
74254                                           No title       en   

                                                 Authors  \
0                           Jefferson, Thomas, 1743-1826   
1                                          United States   
2          Kennedy, John F. (John Fitzgerald), 1917-1963   
3                            Lincoln, Abraham, 1809-1865   
4                                          United States   
...                                                  ...   
74250                             Lee, Vernon, 1856-1935   
74251                         Brewster, David, 1781-1868   
74252  Verne, Jules, 1828-1905; Murphy, Henry Cruse, ...   
74253                               Levy, Amy, 1861-1889   
74254                                                NaN   

                                                Subjects      LoCC  \
0      United States -- History -- Revolution, 1775-1...  E201; JK   
1      Civil rights -- United States -- Sources; Unit...    JK; KF   
2      United States -- Foreign relations -- 1961-196...      E838   
3      Consecration of cemeteries -- Pennsylvania -- ...      E456   
4      United States -- Politics and government -- 17...    JK; KF   
...                                                  ...       ...   
74250                                                NaN       NaN   
74251                                                NaN       NaN   
74252                                                NaN       NaN   
74253                                                NaN       NaN   
74254                                                NaN       NaN   

                                             Bookshelves  
0      Politics; American Revolutionary War; United S...  
1      Politics; American Revolutionary War; United S...  
2       Browsing: History - American; Browsing: Politics  
3      US Civil War; Browsing: History - American; Br...  
4      United States; Politics; American Revolutionar...  
...                                                  ...  
74250                                                NaN  
74251                                                NaN  
74252                                                NaN  
74253                                                NaN  
74254                                                NaN  

[74255 rows x 9 columns]

**Note** all we did here is load the file into the output screen of Jupyter Notebook, we have not actually stored it as a variable we can work with.

We can store it by having the result of the function equal a variable.

In [98]:
df = pd.read_csv("pg_catalog.csv")

The csv file has now been stored as a `DataFrame`. A dataframe is a lot like a spreadsheet with columns and rows, but it has some features to optimize it for data science analysis.

We can show the content of the data frame by type `df`. </br>**Note** you do not have to use the name `df`, this is just a convention.

### 1.2 Loading a Remote File

Loading files from a local copy is easy enough, but it does involve extra steps of going to the website and downloading the file. Pandas allows you to skip this file by simply entering the url of the file you want to download. As entering the url can lead to cluttered code, it is best practice to save that as a string variable first.

In [40]:
gutenberg_csv_url = "https://www.gutenberg.org/cache/epub/feeds/pg_catalog.csv.gz"

In [142]:
df_remote = pd.read_csv(gutenberg_csv_url)

A nice feature of Pandas is that it automatically recognizes common compressed file formats like the `.gz` extension and decompresses them.

#### Critical Thinking: local vs. remote?

When doing any data science project one of the key questions you will have to consider is how you are loading the data for your project. 

> Is it going to be a local copy or a remote connection?

The right answer depends on the use case, but generally what could be some pros and cons of local and remote data?

## 2 Cleaning the DataFrame

Before you start doing anything with your data, you want to make sure you get the DataFrame cleaned up. This will make working with the data easier. 


### 2.1 DataTypes

Part of the power of Pandas is that it assumes that each column of a dataframe is of the same type. It allows it to make calculations faster. When you import from a csv, Pandas is not always able to determine what type of variable is inside the dataframe. We can check how Pandas did by using the `dtypes()` function.

In [143]:
df_remote.dtypes

Text#           int64
Type           object
Issued         object
Title          object
Language       object
Authors        object
Subjects       object
LoCC           object
Bookshelves    object
dtype: object

Pandas was able to figure out that the first column was an integer, but it had trouble with the other columns. It saved these as generic `objects`, which means it can be a `string`, `number` , or other types of data formats. This can lead to problems down the road, and Pandas recommends converting these to their own special strings called `StringDType`. This makes for better storage and processing.

To change one column we can use the following logic:

```python
df_remote['Title']=df_remote['Title'].astype(pd.StringDtype())
```
This basically means this: Take the title column `df_remote['Title']` convert the type `.astype` , set the type to `pd.StringDtype`.


Converting each invidividual column is a bit tedious. Since we know that only the first column is correct and all the other columns need to be `StringDtype` we can create a list of column names and apply the changes to that entire range.

```python
cols = df_remote.columns[1:]
df_remote[cols] = df_remote[cols].astype(pd.StringDtype())
```


In [144]:
#create the list of column names
cols = df_remote.columns[1:]
cols

Index(['Type', 'Issued', 'Title', 'Language', 'Authors', 'Subjects', 'LoCC',
       'Bookshelves'],
      dtype='object')

In [145]:
df_remote[cols] = df_remote[cols].astype(pd.StringDtype())

Now when I check the `dtypes` we should have all strings. 

In [146]:
df_remote.dtypes

Text#                   int64
Type           string[python]
Issued         string[python]
Title          string[python]
Language       string[python]
Authors        string[python]
Subjects       string[python]
LoCC           string[python]
Bookshelves    string[python]
dtype: object

This is unfortunately a very tedious part of the process, but it can prevent a lot of irritation down the road.

### 2.2 Column Names

Another common issue is poorly formatted column names and making these consistent and usuable will reduce errors down the road.
- Pandas generally likes column names that **do not** include spaces or other special characters.
- use only one lower case for column names (this is purely my own preference!)

#### 2.2.1 Remove special characters

One potential issue is that the first column is called `Text#`, the `#` symbol is a special character that could cause confusion. We want to get rid of it. We can use the `.rename()` method to do so. This works in the following way:

```python
df.rename(columns={"Old Column Name":"New Column Name"})
```

In [176]:
df_remote = df_remote.rename(columns={"Text#": "Text_ID"})

In [177]:
df_remote

text_id  type      issued  \
0            1  Text  1971-12-01   
1            2  Text  1972-12-01   
2            3  Text  1973-11-01   
3            4  Text  1973-11-01   
4            5  Text  1975-12-01   
...        ...   ...         ...   
74250    74416  Text  2024-09-14   
74251    74417  Text  2024-09-14   
74252    74418  Text  2024-09-15   
74253    74419  Text  2024-09-15   
74254    90907  Text  2019-12-12   

                                                   title language  \
0      The Declaration of Independence of the United ...       en   
1                       The United States Bill of Rights       en   
2                    John F. Kennedy's Inaugural Address       en   
3                           Lincoln's Gettysburg Address       en   
4                         The United States Constitution       en   
...                                                  ...      ...   
74250                                            Proteus       en   
74251                                   The kaleidoscope       en   
74252                                  Their island home       en   
74253                             Reuben Sachs; a sketch       en   
74254                                           No title       en   

                                                 authors  \
0                           Jefferson, Thomas, 1743-1826   
1                                          United States   
2          Kennedy, John F. (John Fitzgerald), 1917-1963   
3                            Lincoln, Abraham, 1809-1865   
4                                          United States   
...                                                  ...   
74250                             Lee, Vernon, 1856-1935   
74251                         Brewster, David, 1781-1868   
74252  Verne, Jules, 1828-1905; Murphy, Henry Cruse, ...   
74253                               Levy, Amy, 1861-1889   
74254                                               <NA>   

                                                subjects      locc  \
0      United States -- History -- Revolution, 1775-1...  E201; JK   
1      Civil rights -- United States -- Sources; Unit...    JK; KF   
2      United States -- Foreign relations -- 1961-196...      E838   
3      Consecration of cemeteries -- Pennsylvania -- ...      E456   
4      United States -- Politics and government -- 17...    JK; KF   
...                                                  ...       ...   
74250                                               <NA>      <NA>   
74251                                               <NA>      <NA>   
74252                                               <NA>      <NA>   
74253                                               <NA>      <NA>   
74254                                               <NA>      <NA>   

                                             bookshelves  \
0      Politics; American Revolutionary War; United S...   
1      Politics; American Revolutionary War; United S...   
2       Browsing: History - American; Browsing: Politics   
3      US Civil War; Browsing: History - American; Br...   
4      United States; Politics; American Revolutionar...   
...                                                  ...   
74250                                               <NA>   
74251                                               <NA>   
74252                                               <NA>   
74253                                               <NA>   
74254                                               <NA>   

                                        first_author  \
0                       Jefferson, Thomas, 1743-1826   
1                                      United States   
2      Kennedy, John F. (John Fitzgerald), 1917-1963   
3                        Lincoln, Abraham, 1809-1865   
4                                      United States   
...                                              ...   
74250                         Lee, Vernon, 1856-1935   
74251         

#### 2.2.2 Make all column names lower case

We can also rename the column names as a group by passing a function into the rename method and indicating what axis we want to change. In this case, we are going to convert them all to lower case. This prevents you from having to remember to press shift every time you enter the title of a column.

```python
df.rename(str.lower, axis='columns')
```

In [178]:
df_remote = df_remote.rename(str.lower, axis='columns')

In [179]:
df_remote

text_id  type      issued  \
0            1  Text  1971-12-01   
1            2  Text  1972-12-01   
2            3  Text  1973-11-01   
3            4  Text  1973-11-01   
4            5  Text  1975-12-01   
...        ...   ...         ...   
74250    74416  Text  2024-09-14   
74251    74417  Text  2024-09-14   
74252    74418  Text  2024-09-15   
74253    74419  Text  2024-09-15   
74254    90907  Text  2019-12-12   

                                                   title language  \
0      The Declaration of Independence of the United ...       en   
1                       The United States Bill of Rights       en   
2                    John F. Kennedy's Inaugural Address       en   
3                           Lincoln's Gettysburg Address       en   
4                         The United States Constitution       en   
...                                                  ...      ...   
74250                                            Proteus       en   
74251                                   The kaleidoscope       en   
74252                                  Their island home       en   
74253                             Reuben Sachs; a sketch       en   
74254                                           No title       en   

                                                 authors  \
0                           Jefferson, Thomas, 1743-1826   
1                                          United States   
2          Kennedy, John F. (John Fitzgerald), 1917-1963   
3                            Lincoln, Abraham, 1809-1865   
4                                          United States   
...                                                  ...   
74250                             Lee, Vernon, 1856-1935   
74251                         Brewster, David, 1781-1868   
74252  Verne, Jules, 1828-1905; Murphy, Henry Cruse, ...   
74253                               Levy, Amy, 1861-1889   
74254                                               <NA>   

                                                subjects      locc  \
0      United States -- History -- Revolution, 1775-1...  E201; JK   
1      Civil rights -- United States -- Sources; Unit...    JK; KF   
2      United States -- Foreign relations -- 1961-196...      E838   
3      Consecration of cemeteries -- Pennsylvania -- ...      E456   
4      United States -- Politics and government -- 17...    JK; KF   
...                                                  ...       ...   
74250                                               <NA>      <NA>   
74251                                               <NA>      <NA>   
74252                                               <NA>      <NA>   
74253                                               <NA>      <NA>   
74254                                               <NA>      <NA>   

                                             bookshelves  \
0      Politics; American Revolutionary War; United S...   
1      Politics; American Revolutionary War; United S...   
2       Browsing: History - American; Browsing: Politics   
3      US Civil War; Browsing: History - American; Br...   
4      United States; Politics; American Revolutionar...   
...                                                  ...   
74250                                               <NA>   
74251                                               <NA>   
74252                                               <NA>   
74253                                               <NA>   
74254                                               <NA>   

                                        first_author  \
0                       Jefferson, Thomas, 1743-1826   
1                                      United States   
2      Kennedy, John F. (John Fitzgerald), 1917-1963   
3                        Lincoln, Abraham, 1809-1865   
4                                      United States   
...                                              ...   
74250                         Lee, Vernon, 1856-1935   
74251         

## 3 Cleaning the Data

Once the dataframe is in order you will want to clean up some of the data. This is usually a recursive process. That is, you usually only figure out that there is an issue with the data when you start working on it. As you keep finding issues, you want to clean these issues earlier in your code, rather than when you run into them.

### 3.1 Removing formatting codes

When working with string data one common problem is that sometimes formatting codes or special characters are left in the text. For example, the code `\n` is used to indicate a new line in a text. The title for the *Bill of Rights* in this dataframe has a a subtitle. We can see this by getting the value for that cell. We can use the same list access technique we learned in lesson one to access the fourth column and the second value. The only difference is that we have to provide the column name first.

In [180]:
df_remote['title'][1]
# remember [1] is the second value because the first value is [0]

'The United States Bill of Rights'

**Note** In the result above the title includes `\r` and `\n`. In theory, we could manually remove this by changing that specific cell. As there are thousands of rows, going through and manually fixing all this would be extremely tedious. Instead, we'll simply tell Pandas to remove all instances of either `\r` or `\n\` in the column.</br>
There are a couple of ways to do this, but we will simply go through and drop the new line or return character and everything that follows it. We do not need the subtitle.

The code for this is:
```python
df_remote['title'] = df_remote['title'].str.split('\n|\r').str[0]
```
This means take the column `title` (`df_remote['title']`) and split it in two whenever you run into and `\n` or an `\r` (`str.split`). `str.split` actually produces a list of the result `["The United States Bill of Rights", "The Ten Original Amendments to the Constitution of the United States"]`.

In [181]:
df_remote['title'] = df_remote['title'].str.split('\n|\r').str[0]

In [182]:
df_remote

text_id  type      issued  \
0            1  Text  1971-12-01   
1            2  Text  1972-12-01   
2            3  Text  1973-11-01   
3            4  Text  1973-11-01   
4            5  Text  1975-12-01   
...        ...   ...         ...   
74250    74416  Text  2024-09-14   
74251    74417  Text  2024-09-14   
74252    74418  Text  2024-09-15   
74253    74419  Text  2024-09-15   
74254    90907  Text  2019-12-12   

                                                   title language  \
0      The Declaration of Independence of the United ...       en   
1                       The United States Bill of Rights       en   
2                    John F. Kennedy's Inaugural Address       en   
3                           Lincoln's Gettysburg Address       en   
4                         The United States Constitution       en   
...                                                  ...      ...   
74250                                            Proteus       en   
74251                                   The kaleidoscope       en   
74252                                  Their island home       en   
74253                             Reuben Sachs; a sketch       en   
74254                                           No title       en   

                                                 authors  \
0                           Jefferson, Thomas, 1743-1826   
1                                          United States   
2          Kennedy, John F. (John Fitzgerald), 1917-1963   
3                            Lincoln, Abraham, 1809-1865   
4                                          United States   
...                                                  ...   
74250                             Lee, Vernon, 1856-1935   
74251                         Brewster, David, 1781-1868   
74252  Verne, Jules, 1828-1905; Murphy, Henry Cruse, ...   
74253                               Levy, Amy, 1861-1889   
74254                                               <NA>   

                                                subjects      locc  \
0      United States -- History -- Revolution, 1775-1...  E201; JK   
1      Civil rights -- United States -- Sources; Unit...    JK; KF   
2      United States -- Foreign relations -- 1961-196...      E838   
3      Consecration of cemeteries -- Pennsylvania -- ...      E456   
4      United States -- Politics and government -- 17...    JK; KF   
...                                                  ...       ...   
74250                                               <NA>      <NA>   
74251                                               <NA>      <NA>   
74252                                               <NA>      <NA>   
74253                                               <NA>      <NA>   
74254                                               <NA>      <NA>   

                                             bookshelves  \
0      Politics; American Revolutionary War; United S...   
1      Politics; American Revolutionary War; United S...   
2       Browsing: History - American; Browsing: Politics   
3      US Civil War; Browsing: History - American; Br...   
4      United States; Politics; American Revolutionar...   
...                                                  ...   
74250                                               <NA>   
74251                                               <NA>   
74252                                               <NA>   
74253                                               <NA>   
74254                                               <NA>   

                                        first_author  \
0                       Jefferson, Thomas, 1743-1826   
1                                      United States   
2      Kennedy, John F. (John Fitzgerald), 1917-1963   
3                        Lincoln, Abraham, 1809-1865   
4                                      United States   
...                                              ...   
74250                         Lee, Vernon, 1856-1935   
74251         

As you work with a DataFrame there are always other formatting quirks in the date you will want to take care of. It makes little sense to try to clean everything in advance and hope for the best. Likely, you'll find problems as you go and then make the fixes part of the cleaning process.

### Creating New Columns/Feature Engineering

The way data is stored in a spreadsheet "in the wild" can vary drastically. The Gutenberg spreadsheet is pretty well organized, but it stores the author name and vital data (birth and death year) all in the same column. This can be difficult to work with. For example, if I want to order the authors by birth year that data is locked into the author names. We want to get that data out and also follow the standard naming convention of having a column for first name and last name, which makes sorting and searching a lot faster.

Since the names all follow a regular pattern: Last Name, First Name, Years, we can split the column into three different columns using the same `str.split` function used above. By default, `expand` is set to `False`. If we set it to `True`, the result will not be a list, but instead more columns. We can determine the total number of splits by using `n=2`. That is, do this for two commas.

Before we actually change the table we want to test out the result.



In [183]:
df_remote['authors'].str.split(', ', n=2, expand=True)

0                          1  \
0          Jefferson                     Thomas   
1      United States                       <NA>   
2            Kennedy  John F. (John Fitzgerald)   
3            Lincoln                    Abraham   
4      United States                       <NA>   
...              ...                        ...   
74250            Lee                     Vernon   
74251       Brewster                      David   
74252          Verne                      Jules   
74253           Levy                        Amy   
74254           <NA>                       <NA>   

                                                       2  
0                                              1743-1826  
1                                                   <NA>  
2                                              1917-1963  
3                                              1809-1865  
4                                                   <NA>  
...                                                  ...  
74250                                          1856-1935  
74251                                          1781-1868  
74252  1828-1905; Murphy, Henry Cruse, 1810-1882 [Ill...  
74253                                          1861-1889  
74254                                               <NA>  

[74255 rows x 3 columns]

This looked liked it worked, but it created an issue for row 74252. Since there multiple authors for this text the function splits the string by `,` but then lumps all the years together. Although this secondary author is not very interesting, we also don't want to delete it just yet in case we need it later. Instead, we'll save this for now by using the same `str.split` function, but applying it to authors for semi-colons `;` and set the number to `n=1` for one column.

Let's test it out before we change the DataFrame.

In [161]:
df_remote['authors'].str.split('; ', n=1, expand=True)

0  \
0                       Jefferson, Thomas, 1743-1826   
1                                      United States   
2      Kennedy, John F. (John Fitzgerald), 1917-1963   
3                        Lincoln, Abraham, 1809-1865   
4                                      United States   
...                                              ...   
74250                         Lee, Vernon, 1856-1935   
74251                     Brewster, David, 1781-1868   
74252                        Verne, Jules, 1828-1905   
74253                           Levy, Amy, 1861-1889   
74254                                           <NA>   

                                                       1  
0                                                   <NA>  
1                                                   <NA>  
2                                                   <NA>  
3                                                   <NA>  
4                                                   <NA>  
...                                                  ...  
74250                                               <NA>  
74251                                               <NA>  
74252  Murphy, Henry Cruse, 1810-1882 [Illustrator]; ...  
74253                                               <NA>  
74254                                               <NA>  

[74255 rows x 2 columns]

This looks good. Let's create a new column called `'second_author'` and store it there.

In [162]:
df_remote

text_id  type      issued  \
0            1  Text  1971-12-01   
1            2  Text  1972-12-01   
2            3  Text  1973-11-01   
3            4  Text  1973-11-01   
4            5  Text  1975-12-01   
...        ...   ...         ...   
74250    74416  Text  2024-09-14   
74251    74417  Text  2024-09-14   
74252    74418  Text  2024-09-15   
74253    74419  Text  2024-09-15   
74254    90907  Text  2019-12-12   

                                                   title language  \
0      The Declaration of Independence of the United ...       en   
1                       The United States Bill of Rights       en   
2                    John F. Kennedy's Inaugural Address       en   
3                           Lincoln's Gettysburg Address       en   
4                         The United States Constitution       en   
...                                                  ...      ...   
74250                                            Proteus       en   
74251                                   The kaleidoscope       en   
74252                                  Their island home       en   
74253                             Reuben Sachs; a sketch       en   
74254                                           No title       en   

                                                 authors  \
0                           Jefferson, Thomas, 1743-1826   
1                                          United States   
2          Kennedy, John F. (John Fitzgerald), 1917-1963   
3                            Lincoln, Abraham, 1809-1865   
4                                          United States   
...                                                  ...   
74250                             Lee, Vernon, 1856-1935   
74251                         Brewster, David, 1781-1868   
74252  Verne, Jules, 1828-1905; Murphy, Henry Cruse, ...   
74253                               Levy, Amy, 1861-1889   
74254                                               <NA>   

                                                subjects      locc  \
0      United States -- History -- Revolution, 1775-1...  E201; JK   
1      Civil rights -- United States -- Sources; Unit...    JK; KF   
2      United States -- Foreign relations -- 1961-196...      E838   
3      Consecration of cemeteries -- Pennsylvania -- ...      E456   
4      United States -- Politics and government -- 17...    JK; KF   
...                                                  ...       ...   
74250                                               <NA>      <NA>   
74251                                               <NA>      <NA>   
74252                                               <NA>      <NA>   
74253                                               <NA>      <NA>   
74254                                               <NA>      <NA>   

                                             bookshelves  
0      Politics; American Revolutionary War; United S...  
1      Politics; American Revolutionary War; United S...  
2       Browsing: History - American; Browsing: Politics  
3      US Civil War; Browsing: History - American; Br...  
4      United States; Politics; American Revolutionar...  
...                                                  ...  
74250                                               <NA>  
74251                                               <NA>  
74252                                               <NA>  
74253                                               <NA>  
74254                                               <NA>  

[74255 rows x 9 columns]

In [164]:
df_remote[['first_author','second_author']]=df_remote['authors'].str.split('; ', n=1, expand=True)

In [165]:
df_remote

text_id  type      issued  \
0            1  Text  1971-12-01   
1            2  Text  1972-12-01   
2            3  Text  1973-11-01   
3            4  Text  1973-11-01   
4            5  Text  1975-12-01   
...        ...   ...         ...   
74250    74416  Text  2024-09-14   
74251    74417  Text  2024-09-14   
74252    74418  Text  2024-09-15   
74253    74419  Text  2024-09-15   
74254    90907  Text  2019-12-12   

                                                   title language  \
0      The Declaration of Independence of the United ...       en   
1                       The United States Bill of Rights       en   
2                    John F. Kennedy's Inaugural Address       en   
3                           Lincoln's Gettysburg Address       en   
4                         The United States Constitution       en   
...                                                  ...      ...   
74250                                            Proteus       en   
74251                                   The kaleidoscope       en   
74252                                  Their island home       en   
74253                             Reuben Sachs; a sketch       en   
74254                                           No title       en   

                                                 authors  \
0                           Jefferson, Thomas, 1743-1826   
1                                          United States   
2          Kennedy, John F. (John Fitzgerald), 1917-1963   
3                            Lincoln, Abraham, 1809-1865   
4                                          United States   
...                                                  ...   
74250                             Lee, Vernon, 1856-1935   
74251                         Brewster, David, 1781-1868   
74252  Verne, Jules, 1828-1905; Murphy, Henry Cruse, ...   
74253                               Levy, Amy, 1861-1889   
74254                                               <NA>   

                                                subjects      locc  \
0      United States -- History -- Revolution, 1775-1...  E201; JK   
1      Civil rights -- United States -- Sources; Unit...    JK; KF   
2      United States -- Foreign relations -- 1961-196...      E838   
3      Consecration of cemeteries -- Pennsylvania -- ...      E456   
4      United States -- Politics and government -- 17...    JK; KF   
...                                                  ...       ...   
74250                                               <NA>      <NA>   
74251                                               <NA>      <NA>   
74252                                               <NA>      <NA>   
74253                                               <NA>      <NA>   
74254                                               <NA>      <NA>   

                                             bookshelves  \
0      Politics; American Revolutionary War; United S...   
1      Politics; American Revolutionary War; United S...   
2       Browsing: History - American; Browsing: Politics   
3      US Civil War; Browsing: History - American; Br...   
4      United States; Politics; American Revolutionar...   
...                                                  ...   
74250                                               <NA>   
74251                                               <NA>   
74252                                               <NA>   
74253                                               <NA>   
74254                                               <NA>   

                                        first_author  \
0                       Jefferson, Thomas, 1743-1826   
1                                      United States   
2      Kennedy, John F. (John Fitzgerald), 1917-1963   
3                        Lincoln, Abraham, 1809-1865   
4                                      United States   
...                                              ...   
74250                         Lee, Vernon, 1856-1935   
74251         

Now that that the authors have been split, we can split up the first author name by 'first_name','last_name', and 'years'.

In [166]:
df_remote[['last_name', 'first_name','years']] = df_remote['authors'].str.split(', ', n=2, expand=True)

In [167]:
df_remote

text_id  type      issued  \
0            1  Text  1971-12-01   
1            2  Text  1972-12-01   
2            3  Text  1973-11-01   
3            4  Text  1973-11-01   
4            5  Text  1975-12-01   
...        ...   ...         ...   
74250    74416  Text  2024-09-14   
74251    74417  Text  2024-09-14   
74252    74418  Text  2024-09-15   
74253    74419  Text  2024-09-15   
74254    90907  Text  2019-12-12   

                                                   title language  \
0      The Declaration of Independence of the United ...       en   
1                       The United States Bill of Rights       en   
2                    John F. Kennedy's Inaugural Address       en   
3                           Lincoln's Gettysburg Address       en   
4                         The United States Constitution       en   
...                                                  ...      ...   
74250                                            Proteus       en   
74251                                   The kaleidoscope       en   
74252                                  Their island home       en   
74253                             Reuben Sachs; a sketch       en   
74254                                           No title       en   

                                                 authors  \
0                           Jefferson, Thomas, 1743-1826   
1                                          United States   
2          Kennedy, John F. (John Fitzgerald), 1917-1963   
3                            Lincoln, Abraham, 1809-1865   
4                                          United States   
...                                                  ...   
74250                             Lee, Vernon, 1856-1935   
74251                         Brewster, David, 1781-1868   
74252  Verne, Jules, 1828-1905; Murphy, Henry Cruse, ...   
74253                               Levy, Amy, 1861-1889   
74254                                               <NA>   

                                                subjects      locc  \
0      United States -- History -- Revolution, 1775-1...  E201; JK   
1      Civil rights -- United States -- Sources; Unit...    JK; KF   
2      United States -- Foreign relations -- 1961-196...      E838   
3      Consecration of cemeteries -- Pennsylvania -- ...      E456   
4      United States -- Politics and government -- 17...    JK; KF   
...                                                  ...       ...   
74250                                               <NA>      <NA>   
74251                                               <NA>      <NA>   
74252                                               <NA>      <NA>   
74253                                               <NA>      <NA>   
74254                                               <NA>      <NA>   

                                             bookshelves  \
0      Politics; American Revolutionary War; United S...   
1      Politics; American Revolutionary War; United S...   
2       Browsing: History - American; Browsing: Politics   
3      US Civil War; Browsing: History - American; Br...   
4      United States; Politics; American Revolutionar...   
...                                                  ...   
74250                                               <NA>   
74251                                               <NA>   
74252                                               <NA>   
74253                                               <NA>   
74254                                               <NA>   

                                        first_author  \
0                       Jefferson, Thomas, 1743-1826   
1                                      United States   
2      Kennedy, John F. (John Fitzgerald), 1917-1963   
3                        Lincoln, Abraham, 1809-1865   
4                                      United States   
...                                              ...   
74250                         Lee, Vernon, 1856-1935   
74251         

We can then further split `years` by birth `year` and `death` year using `-` as the character regular expression.

In [169]:
df_remote[['birth', 'death']] = df_remote['years'].str.split('-', n=1, expand=True)

In [184]:
df_remote

text_id  type      issued  \
0            1  Text  1971-12-01   
1            2  Text  1972-12-01   
2            3  Text  1973-11-01   
3            4  Text  1973-11-01   
4            5  Text  1975-12-01   
...        ...   ...         ...   
74250    74416  Text  2024-09-14   
74251    74417  Text  2024-09-14   
74252    74418  Text  2024-09-15   
74253    74419  Text  2024-09-15   
74254    90907  Text  2019-12-12   

                                                   title language  \
0      The Declaration of Independence of the United ...       en   
1                       The United States Bill of Rights       en   
2                    John F. Kennedy's Inaugural Address       en   
3                           Lincoln's Gettysburg Address       en   
4                         The United States Constitution       en   
...                                                  ...      ...   
74250                                            Proteus       en   
74251                                   The kaleidoscope       en   
74252                                  Their island home       en   
74253                             Reuben Sachs; a sketch       en   
74254                                           No title       en   

                                                 authors  \
0                           Jefferson, Thomas, 1743-1826   
1                                          United States   
2          Kennedy, John F. (John Fitzgerald), 1917-1963   
3                            Lincoln, Abraham, 1809-1865   
4                                          United States   
...                                                  ...   
74250                             Lee, Vernon, 1856-1935   
74251                         Brewster, David, 1781-1868   
74252  Verne, Jules, 1828-1905; Murphy, Henry Cruse, ...   
74253                               Levy, Amy, 1861-1889   
74254                                               <NA>   

                                                subjects      locc  \
0      United States -- History -- Revolution, 1775-1...  E201; JK   
1      Civil rights -- United States -- Sources; Unit...    JK; KF   
2      United States -- Foreign relations -- 1961-196...      E838   
3      Consecration of cemeteries -- Pennsylvania -- ...      E456   
4      United States -- Politics and government -- 17...    JK; KF   
...                                                  ...       ...   
74250                                               <NA>      <NA>   
74251                                               <NA>      <NA>   
74252                                               <NA>      <NA>   
74253                                               <NA>      <NA>   
74254                                               <NA>      <NA>   

                                             bookshelves  \
0      Politics; American Revolutionary War; United S...   
1      Politics; American Revolutionary War; United S...   
2       Browsing: History - American; Browsing: Politics   
3      US Civil War; Browsing: History - American; Br...   
4      United States; Politics; American Revolutionar...   
...                                                  ...   
74250                                               <NA>   
74251                                               <NA>   
74252                                               <NA>   
74253                                               <NA>   
74254                                               <NA>   

                                        first_author  \
0                       Jefferson, Thomas, 1743-1826   
1                                      United States   
2      Kennedy, John F. (John Fitzgerald), 1917-1963   
3                        Lincoln, Abraham, 1809-1865   
4                                      United States   
...                                              ...   
74250                         Lee, Vernon, 1856-1935   
74251         

All of these changes leave us with several columns that contain old data, `authors`, `first_author`, `years`. We can remove these using the `.drop()` method on the dataframe. We have to specify the columns we want to drop using `columns=`.

In [185]:
df_remote.drop(columns=['authors','first_author','years'])

text_id  type      issued  \
0            1  Text  1971-12-01   
1            2  Text  1972-12-01   
2            3  Text  1973-11-01   
3            4  Text  1973-11-01   
4            5  Text  1975-12-01   
...        ...   ...         ...   
74250    74416  Text  2024-09-14   
74251    74417  Text  2024-09-14   
74252    74418  Text  2024-09-15   
74253    74419  Text  2024-09-15   
74254    90907  Text  2019-12-12   

                                                   title language  \
0      The Declaration of Independence of the United ...       en   
1                       The United States Bill of Rights       en   
2                    John F. Kennedy's Inaugural Address       en   
3                           Lincoln's Gettysburg Address       en   
4                         The United States Constitution       en   
...                                                  ...      ...   
74250                                            Proteus       en   
74251                                   The kaleidoscope       en   
74252                                  Their island home       en   
74253                             Reuben Sachs; a sketch       en   
74254                                           No title       en   

                                                subjects      locc  \
0      United States -- History -- Revolution, 1775-1...  E201; JK   
1      Civil rights -- United States -- Sources; Unit...    JK; KF   
2      United States -- Foreign relations -- 1961-196...      E838   
3      Consecration of cemeteries -- Pennsylvania -- ...      E456   
4      United States -- Politics and government -- 17...    JK; KF   
...                                                  ...       ...   
74250                                               <NA>      <NA>   
74251                                               <NA>      <NA>   
74252                                               <NA>      <NA>   
74253                                               <NA>      <NA>   
74254                                               <NA>      <NA>   

                                             bookshelves  \
0      Politics; American Revolutionary War; United S...   
1      Politics; American Revolutionary War; United S...   
2       Browsing: History - American; Browsing: Politics   
3      US Civil War; Browsing: History - American; Br...   
4      United States; Politics; American Revolutionar...   
...                                                  ...   
74250                                               <NA>   
74251                                               <NA>   
74252                                               <NA>   
74253                                               <NA>   
74254                                               <NA>   

                                           second_author      last_name  \
0                                                   <NA>      Jefferson   
1                                                   <NA>  United States   
2                                                   <NA>        Kennedy   
3                                                   <NA>        Lincoln   
4                                                   <NA>  United States   
...                                                  ...            ...   
74250                                               <NA>            Lee   
74251                                               <NA>       Brewster   
74252  Murphy, Henry Cruse, 1810-1882 [Illustrator]; ...          Verne   
74253                                               <NA>           Levy   
74254                                               <NA>           <NA>   

                      first_name birth  \
0                         Thomas  1743   
1                           <NA>  <NA>   
2      John F. (John Fitzgerald)  1917   
3                        Abraham  1809   
4                           <NA>  <NA>   
...                          .

The above previous shows the desired result. We can make these changes permanent by saving this into df_remote dataframe.

In [187]:
df_remote = df_remote.drop(columns=['authors','first_author','years'])

## 4 Saving Progress

For now, we have a DataFrame we can work with although undoubtedly we'll run into unforeseen issues down the road. It is a good practice to save the dataframe. This prevents us from having to run the above code everytime. There are a couple of ways we can save this. We can use the pandas function `write_csv()`, which converts this to a csv file. The advantage of a CSV file is that any computer can read them.
</br> 
There are several issues with CSV files:
- They tend to be big
- Read and write times can be slow
- Will add an empty index column if you are not careful
- Unless you specifically indicate the column types, the dtype will get lost. This is a huge pain.

Alternatively, we can also save this to a `.pickle` file.

**Advantages**

- Smaller
- Faster
- Keeps dtypes

**Disadvantages**

- Requires Python to open
  
  

Saving to pickle file is incredibly simple. 

df_remote.to_pickle('pg_catalog_clean.pickle')

The code below should place the file in your working directory.

### 4.1 Testing the Different File Types (optional)

If you are really curious about the difference between `.csv` files and `.pickle` files. The code below shows you both the difference in load time and file size for each. You will also see that the CSV file import caused an extra column to appear `Unnamed:0` and convert all the dtypes back to object.

In [189]:
df_remote.to_csv('pg_catalog_clean.csv')

In [200]:
%timeit df_clean_pg_csv = pd.read_csv('pg_catalog_clean.csv')
df_clean_pg_csv.dtypes

304 ms ± 10.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)


Unnamed: 0        int64
text_id           int64
type             object
issued           object
title            object
language         object
subjects         object
locc             object
bookshelves      object
second_author    object
last_name        object
first_name       object
birth            object
death            object
dtype: object

In [199]:
%timeit df_clean_pg_pickle = pd.read_pickle('pg_catalog_clean.pickle')
df_clean_pg_pickle.dtypes

82.8 ms ± 6.72 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)


text_id                   int64
type             string[python]
issued           string[python]
title                    object
language         string[python]
subjects         string[python]
locc             string[python]
bookshelves      string[python]
second_author    string[python]
last_name        string[python]
first_name       string[python]
birth            string[python]
death            string[python]
dtype: object

We can also check how large those files are in memory by calling up the info on the current working directory. We'll note the `csv` files are around 20mb and the `pickle` file is around 15mb.

In [216]:
%ls -Gflash *

 Volume in drive C is Ptolemy
 Volume Serial Number is 7089-6643

 Directory of C:\Users\joost\My Drive (joostburgers@gmail.com)\Teaching\JMU\Courses\digital_studies_resources\repos\python_lessons\lesson_2


 Directory of C:\Users\joost\My Drive (joostburgers@gmail.com)\Teaching\JMU\Courses\digital_studies_resources\repos\python_lessons\lesson_2

20-Sep-24  02:22 PM    <DIR>          .
20-Sep-24  02:22 PM    <DIR>          ..
19-Sep-24  09:01 PM    <DIR>          .ipynb_checkpoints
17-Sep-24  04:36 PM    <DIR>          data
20-Sep-24  02:22 PM           188,053 lesson_2_scraping_gutenberg.ipynb
19-Sep-24  02:51 PM        20,133,367 pg_catalog.csv
20-Sep-24  02:03 PM        20,758,071 pg_catalog_clean.csv
20-Sep-24  02:01 PM        15,325,990 pg_catalog_clean.pickle
               4 File(s)     56,405,481 bytes
               4 Dir(s)  11,161,288,704 bytes free


In [16]:
virginia_df = df[df['subjects'].str.contains('Virginia -- History', case=False, na=False) & 
                 df['subjects'].str.contains('History', case=False, na=False)]

In [21]:
virginia_df['title'].head(50)

72       The Red Badge of Courage: An Episode of the Am...
444      The Red Badge of Courage: An Episode of the Am...
2767     With Lee in Virginia: A Story of the American ...
2769                                   To Have and to Hold
2858     Pioneers of the Old South: A Chronicle of Engl...
4206     A Briefe and True Report of the New Found Land...
4487     The Heart's Highway: A Romance of Virginia in ...
7423     Richard of Jamestown : a Story of the Virginia...
8080                                        The Virginians
8381     Mohun; Or, the Last Days of Lee and His Paladins.
9975                                 Salute to Adventurers
10023    A Soldier of Virginia: A Tale of Colonel Washi...
14442                                               Audrey
19083    With Lee in Virginia: A Story of the American ...
21411                     The Settlers: A Tale of Virginia
21815       Prisoners of Hope: A Tale of Colonial Virginia
21996     The Story of a Cannoneer Under Stonewall Jacks

In [18]:
virginia_df.dtypes


text#           int64
type           object
issued         object
title          object
language       object
subjects       object
locc           object
bookshelves    object
last_name      object
first_name     object
birth          object
death          object
dtype: object

In [22]:
GUTENBERG_ROBOT_URL = "http://www.gutenberg.org/robot/harvest?filetypes[]=txt"
r = requests.get(GUTENBERG_ROBOT_URL)


In [23]:
GUTENBERG_MIRROR = re.search('(https?://[^/]+)[^"]*.zip', r.text).group(1)
GUTENBERG_MIRROR

'http://aleph.gutenberg.org'

In [24]:
def gutenberg_text_urls(id: str, mirror=GUTENBERG_MIRROR, suffixes=("", "-8", "-0")) -> list[str]:
    path = "/".join(id[:-1]) or "0"
    return [f"{mirror}/{path}/{id}/{id}{suffix}.zip" for suffix in suffixes]

In [25]:
def download_gutenberg(id: str) -> str:
    for url in gutenberg_text_urls(id):
        r = requests.get(url)
        if r.status_code == 404:
            logging.warning(f"404 for {url}")
            continue
        r.raise_for_status()
        break
    
    z = zipfile.ZipFile(BytesIO(r.content))
    
    if len(z.namelist()) != 1:
        raise Exception(f"Expected 1 file in {z.namelist()}")
        
    return z.read(z.namelist()[0]).decode('utf-8')

In [26]:
def strip_headers(text):
    GUTENBERG_TEXT = "PROJECT GUTENBERG EBOOK"
    
    in_text = False
    output = []
    
    for line in text.splitlines():        
        if GUTENBERG_TEXT in line:
            if not in_text:
                in_text = True
            else:
                break
        else:
            if in_text:
                output.append(line)

    return "\n".join(output).strip()



In [27]:
GUTENBERG_TEXT_URL = "https://www.gutenberg.org/ebooks/{id}.txt.utf-8"

In [28]:
def book_text(book_id):
    r = requests.get(GUTENBERG_TEXT_URL.format(id=book_id))
    text = r.text
    clean_text = strip_headers(text)
    return clean_text

In [29]:
data_path = Path("data")
data_path.mkdir(exist_ok=True)

In [ ]:
for index, book in tqdm(virginia_df.iterrows(),total=len(virginia_df)):  # Use iterrows() to iterate over rows
    id = book['text#']  # Assuming the second column contains the ID
    text = book_text(id)
   #print(f"Saving {book['title']} by {book['last_name']}, {book['first_name']} containing {len(text):_} characters")

   
    filename = f"{book['last_name']}_{book['first_name']}_{book['title']}"
    filename = re.sub(r'[<>:;"/\\|?*]', '', filename)  # Remove invalid characters
    filename = filename.replace('\n', '')  # Ensure no newlines in filename
    filename = filename[:75]
    filename = filename + ".txt"

    # Specify encoding as 'utf-8' to handle a wider range of characters
    with open(data_path / filename, "wt", encoding='utf-8') as f:
        f.write(text)

In [31]:
def fetch_text_data(df):
    # Create a new column 'text_data' to store the fetched text
    df['text_data'] = ''
    
    # Use itertuples for better performance
    for book in tqdm(df.itertuples(index=True, name=None), total=len(df)):
        # Access columns by index (0 is the index, 1 is the first column, 2 is the second, etc.)
        index = book[0]  # Index of the row
        text_id = book[1]  # Assuming 'text#' is the second column
        text = book_text(text_id)  # Function to fetch the book text
        
        # Insert the fetched text into the new column
        df.at[index, 'text_data'] = text

    return df



In [32]:
virginia_df_local = virginia_df.loc[:]
virginia_df_local_text = fetch_text_data(virginia_df_local)

100%|██████████| 67/67 [00:17<00:00,  3.76it/s]


In [ ]:
virginia_df_local_text.head(10)
